packages/hydrogen/src/product/useOptimisticProduct.ts

```diff
@@ -0,0 +1,104 @@
+import {useNavigation} from '@remix-run/react';
+import {
+  Product,
+  ProductVariant,
+} from '@shopify/hydrogen-react/storefront-api-types';
+import {useEffect, useState} from 'react';
+import type {PartialDeep} from 'type-fest';
+
+type OptimisticProduct<T> = T & {
+  isOptimistic?: boolean;
+};
+
+type OptimisticProductInput = Product & {
+  selectedVariant?: PartialDeep<ProductVariant>;
+};
+
+type ProductWithVariants = {
+  product: {
+    variants: {nodes: Array<PartialDeep<ProductVariant>>};
+  };
+};
+
+type OptimisticProductVariants =
+  | Array<PartialDeep<ProductVariant>>
+  | Promise<Array<PartialDeep<ProductVariant>>>
+  | ProductWithVariants
+  | Promise<ProductWithVariants>;
+
+/**
+ * @param product The product object from `context.storefront.query()` returned by a server loader.
+ * @param variants The available product variants for the product. This can be an array of variants, a promise that resolves to an array of variants, or an object with a `product` key that contains the variants.
+ * @returns A new product object where the `selectedVariant` property is set to the variant that matches the current URL search params. If no variant is found, the original product object is returned. The `isOptimistic` property is set to `true` if the `selectedVariant` has been optimistically changed.
+ */
+export function useOptimisticProduct<
+  ProductWithSelectedVariant = OptimisticProductInput,
+  Variants = OptimisticProductVariants,
+>(
+  product: ProductWithSelectedVariant,
+  variants: Variants,
+): OptimisticProduct<ProductWithSelectedVariant> {
+  const navigation = useNavigation();
+  const [resolvedVariants, setResolvedVariants] = useState<
+    Array<PartialDeep<ProductVariant>>
+  >([]);
+
+  useEffect(() => {
+    Promise.resolve(variants)
+      .then((productWithVariants) => {
+        if (productWithVariants) {
+          setResolvedVariants(
+            productWithVariants instanceof Array
+              ? productWithVariants
+              : (productWithVariants as unknown as ProductWithVariants).product
```

Comment:
Curious, why is the double cast `as unknown as ProductWithVariants` needed here?

Also shouldn't we be more defensive here when accessing deeply nested fields in `.product.variants.node` ?

Reply:
Good catch